doc/Native APIs.md

```diff
@@ -0,0 +1,417 @@
+Native API Documentation
+========================
+
+This documents the APIs that can be used to implement native wrappers. There are
+three types of APIs:
+
+- `student_view_data`: Exposes block settings and content as JSON data. Can be
+  retrieved via the edX Course Blocks API. Does not include user-specific data
+  (i.e. items derived from `Scope.user_state` fields).
+- `student_view_user_state`: XBlock handler which returns user state data in
+  JSON format.
+- Custom XBlock handlers used for submitting user input and recording user actions.
+
+Problem Builder (`problem-builder`)
+-----------------------------------
+
+### `student_view_data`
+
+- `max_attempts`: Max number of allowed attempts (integer).
+- `extended_feedback`: `true` if extended feedback is enabled for this block
+  (boolean).
+- `feedback_label`: Feedback message header (string).
+- `messages`: A dict with three string entries: `completed`, `incomplete`, and
+  `max_attempts_reached`. See below for info on each (object).
+- `components`: A list of `student_view_data` output of all immediate child
+  components which implement `student_view_data`. Child components which don't
+  implement `student_view_data` are omitted from the list (array).
+
+#### `messages`
+
+These messages are displayed as feedback, depending on the state of the block.
+
+- `completed`: Message to be shown if the user successfully completes the
+  problem. May contain HTML (string).
+- `incomplete`: Message to be shown if the user did not successfully complete
+  the problem, but still has available attempts left. May contain HTML (string).
+- `max_attempts_reached`: Message to be shown if the user did not complete the
+  problem and has no available attempts left.
+
+### `student_view_user_state`
```

Comment:
I intentionally omitted fields which I think are not needed. I think `step` and `next_step` are only used by the deprecated assessment mode, and `attempted` is not used anywhere AFAIK and we might want to remove it in  the future, so I don't want API consumers to start depending on it.

`"edx_notes_visibility"` is some edx thing not directly related to problem builder.

I will make a ticket to implement `student_view_user_state` for each block manually instead of exposing all `user_state` fields by default. It's more work, but that way we won't accidentally start exposing internal fields via the API, and have API consumers start depending on them.